pycontrails/datalib/spire.py

```diff
@@ -737,3 +754,663 @@ def _downsample_flight(
+        if "departure_airport_alt_ft" not in row.index:
+            raise ValueError("field departure_airport_alt_ft must be present.")
+
+        departure_lon = row["departure_airport_lon"]
+        departure_lat = row["departure_airport_lat"]
+        departure_alt_ft = row["departure_airport_alt_ft"]
+        if any(
+            [
+                np.isnan(departure_lon),
+                np.isnan(departure_lat),
+                np.isnan(departure_alt_ft),
+            ]
+        ):
+            return np.nan
+
+        return cls._calc_distance_m(
+            lon_0=row["longitude"],
+            lat_0=row["latitude"],
+            alt_ft_0=row["altitude_baro"],
+            lon_f=departure_lon,
+            lat_f=departure_lat,
+            alt_ft_f=departure_alt_ft,
+        )
+
+    @classmethod
+    def _calc_dist_to_arrival_airport(cls, row: pd.Series) -> float:
+        """
+        Calculate the distance from a given waypoint to the arrival airport.
+
+        Returns
+        -------
+        distance in meters to arrival airport.
+        np.nan if it cannot be calculated.
+        """
+
+        if "arrival_airport_lat" not in row.index:
+            raise ValueError("field arrival_airport_lat must be present.")
+        if "arrival_airport_lon" not in row.index:
+            raise ValueError("field arrival_airport_lon must be present.")
+        if "arrival_airport_alt_ft" not in row.index:
+            raise ValueError("field arrival_airport_alt_ft must be present.")
+
+        arrival_lon = row["arrival_airport_lon"]
+        arrival_lat = row["arrival_airport_lat"]
+        arrival_alt_ft = row["arrival_airport_alt_ft"]
+        if any([np.isnan(arrival_lon), np.isnan(arrival_lat), np.isnan(arrival_alt_ft)]):
+            return np.nan
+
+        return cls._calc_distance_m(
+            lon_0=row["longitude"],
+            lat_0=row["latitude"],
+            alt_ft_0=row["altitude_baro"],
+            lon_f=arrival_lon,
+            lat_f=arrival_lat,
+            alt_ft_f=arrival_alt_ft,
+        )
+
+    def _calculate_additional_fields(self):
+        """
+        Add additional columns to the provided dataframe.
+
+        These additional fields are needed to apply the validation ruleset.
+        """
+        self._df = self._df.assign(
+            elapsed_seconds=[
+                self._rolling_time_delta_seconds(window) for window in self._df.rolling(window=2)
+            ],
+        )
+        self._df = self._df.assign(
+            elapsed_distance_m=[
+                self._rolling_distance_meters(window) for window in self._df.rolling(window=2)
+            ],
+        )
+        self._df = self._df.assign(
+            ground_speed_m_s=self._df["elapsed_distance_m"]
+            .divide(self._df["elapsed_seconds"])
+            .replace(np.inf, np.nan)
+        )
+        self._df = self._df.assign(
+            rocd_fps=[self._rolling_rocd_fps(window) for window in self._df.rolling(window=2)]
+        )
```

Comment:
My main piece of feedback for this PR: there are a number of `pycontrails.Flight` methods and other helper functions that already do these computations in a vectorized manner. Some of them may need slight tweaking to match exactly what you're doing here, but it shouldn't be too hard. Is there a reason to avoid them?

If you have a good reason to avoid the existing pycontrails tooling (and there definitely are some good reasons -- some of the existing pycontrails function have opinions / conventions baked into them), all of these calculations can still be vectorized. For example, your calculation for `elapsed_seconds` is essentially:

```
elapsed_seconds = self._df["timestamp"].diff().dt.total_seconds().astype("Int64")
```

Reply:
Great. No reason to have avoided methods that have parity of behavior using a vectorized approach (other than my fairly shallow experience w. pd and np).  I'd like to avoid referencing existing pycontrails methods that have domain logic/assumptions, but otherwise welcome refactor of these to avoid `pd.rolling()` (I assume that will help w/ perf notably).

@zebengberg would you mind helping with that refactor? Feel free to push changes directly to this branch, or throw up small PRs pointed to this branch with the proposed refactor.